modeling/algorithm/local-shared-algorithm.go

```diff
@@ -0,0 +1,205 @@
+// endpoints to zones has no endpoints.
+type LocalSharedSliceAlgorithm struct{}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy. Zones with no endpoints allocated will be treated as a whole
+// that shares a shared-SG.
+func (alg LocalSharedSliceAlgorithm) CreateSliceGroups(region types.RegionInfo) (map[string]types.EndpointSliceGroup, error) {
+	if region.ZoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	// if number of total endpoints < number of zones, use original algorithm
+	// instead. This algorithm itself can handle some of these special corner
+	// cases but performs poorly at small scale corner cases, so using the
+	// original algorithm seems a better solution in terms of performance and
+	// simplicity.
+	if region.TotalEndpoints < len(region.ZoneDetails) {
+		return OriginalAlgorithm{}.CreateSliceGroups(region)
+	}
+	sliceGroups := map[string]types.EndpointSliceGroup{}
+	// endpointsNeeded stores zones with number of endpoints needed
+	endpointsNeeded := endpointsList{}
+	// endpointsNeededUrgent stores high priority zones with number of endpoints
+	// needed
+	endpointsNeededUrgent := endpointsList{}
+	// availablePool is used to contribute endpoints shared with zones when there
+	// are not enough endpoints in the available list.
+	availablePool := ZonePriorityQueue{
+		Region: region,
+	}
+	// receiverPool contains all zones. This is used to receive extra endpoints
+	// when there are not enough endpoints in the needed list.
+	receiverPool := ZonePriorityQueue{
+		Region:          region,
+		ReceiveEndpoint: true,
+	}
+	for zoneName, zone := range region.ZoneDetails {
+		var localGroup types.EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.ZoneTrafficWeights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]types.WeightedEndpoints{}
+
+		// calculate expected number of endpoints based on the proportion of
+		// nodes in this zone
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means need more endpoints from other
+		// sliceGroups, a positive value means need give out endpoints to other
+		// sliceGroups
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// if deviation > 0, this zone is a qualified candidate for
+		// availablePool that contributes endpoints to other zones
+		validCandidate := !math.Signbit(deviation)
+		if validCandidate {
+			availablePool.ZoneNames = append(availablePool.ZoneNames, zoneName)
+		}
+		receiverPool.ZoneNames = append(receiverPool.ZoneNames, zoneName)
+		// merge all the zones have no endpoints into a whole
```

Comment:
Maybe something like this?
```suggestion
		// merge all the zones with no endpoints into a shared slice group
```

Reply:
changed